File: quota-viewer/src/api/quotaOverview.ts
import type {
  AccountQuotaState,
  DailyUsagePoint,
  MembershipStatus,
  QuotaItem,
  QuotaOverview,
  QuotaState,
  WalletState
} from '@/types'

type DecimalString = string

interface QuotaWindowDTO {
  kind: string
  state: 'active' | 'exhausted' | 'unknown'
  anchor_at: string
  period_start: string | null
  period_end: string | null
  resets_at: string | null
  limit: DecimalString | null
  used: DecimalString | null
  remaining: DecimalString | null
  used_percent: number | null
}

export interface QuotaOverviewDTO {
  schema_version: number
  generated_at: string
  as_of: string
  fresh_until: string
  display_timezone: string
  freshness: 'fresh' | 'stale' | 'unknown'
  account: {
    display_label: string
    data_scope: string
    quota_state: AccountQuotaState
  }
  wallet: {
    unit: string
    state: WalletState
    available: DecimalString
    reserved: DecimalString
    balance_billed_key_count: number
    today_spend?: DecimalString | null
    month_spend?: DecimalString | null
  }
  subscriptions: QuotaSubscriptionDTO[]
}

interface QuotaSubscriptionDTO {
  id: string
  group_id: string
  name: string
  status: string
  starts_at: string
  expires_at: string
  weekly_window: QuotaWindowDTO
  monthly_window?: QuotaWindowDTO | null
  period_usage?: {
    state: 'available' | 'unknown'
    observed_until?: string
    bucket_kind?: string
    total_requests?: number | null
    total_tokens?: number | null
    points?: Array<{
      index: number
      start_at: string
      end_at: string
      state: 'complete' | 'partial' | 'future'
      requests: number | null
      cache_hit_tokens: number | null
      cache_miss_tokens: number | null
      output_tokens: number | null
      total_tokens: number | null
    }>
  }
}

const formatDateTime = (value: string | null | undefined) => {
  if (!value) return '—'
  const date = new Date(value)
  if (Number.isNaN(date.getTime())) return '—'
  return new Intl.DateTimeFormat('zh-CN', {
    month: 'numeric',
    day: 'numeric',
    hour: '2-digit',
    minute: '2-digit',
    hour12: false
  }).format(date)
}

const formatUpdatedAt = (value: string) => {
  const date = new Date(value)
  if (Number.isNaN(date.getTime())) return '未知时间'
  return new Intl.DateTimeFormat('zh-CN', {
    hour: '2-digit',
    minute: '2-digit',
    hour12: false
  }).format(date)
}

const validTimestampOrNull = (value: string | null | undefined) => {
  if (!value) return null
  return Number.isFinite(new Date(value).getTime()) ? value : null
}

const formatSnow = (value: DecimalString | null) =>
  value == null ? '—' : `❄${decimalToFixed(value, 2)}`

const decimalToFixed = (value: string, digits: number) => {
  const match = value.trim().match(/^(\d+)(?:\.(\d+))?$/)
  if (!match) return '—'
  const integer = match[1]
  const fraction = match[2] ?? ''
  const padded = `${fraction}${'0'.repeat(digits + 1)}`.slice(0, digits + 1)
  const scale = 10n ** BigInt(digits)
  const base = BigInt(integer) * scale + BigInt(padded.slice(0, digits) || '0')
  const rounded = base + (Number(padded[digits] ?? '0') >= 5 ? 1n : 0n)
  const resultInteger = rounded / scale
  const resultFraction = String(rounded % scale).padStart(digits, '0')
  return digits > 0 ? `${resultInteger}.${resultFraction}` : String(resultInteger)
}

const mapState = (subscription: QuotaSubscriptionDTO): QuotaState => {
  if (subscription.status !== 'active') {
    return ['expired', 'suspended', 'revoked'].includes(subscription.status)
      ? 'expired'
      : 'unknown'
  }

  const monthlyWindow = subscription.monthly_window
  const windows = monthlyWindow
    ? [subscription.weekly_window, monthlyWindow]
    : null
  if (
    !windows ||
    windows.some(
      (window) =>
        window.state === 'unknown' ||
        typeof window.used_percent !== 'number' ||
        !Number.isFinite(window.used_percent)
    )
  ) {
    return 'unknown'
  }
  if (windows.some((window) => window.state === 'exhausted')) return 'exhausted'
  return windows.some((window) => (window.used_percent ?? 0) >= 90)
    ? 'warning'
    : 'available'
}

const mapMembershipStatus = (status: string): MembershipStatus =>
  ['active', 'expired', 'suspended', 'revoked'].includes(status)
    ? (status as MembershipStatus)
    : 'unknown'

const membershipStatusDetail = (
  status: MembershipStatus,
  expiresAt: string | null
) => {
  switch (status) {
    case 'expired': {
      const formatted = formatDateTime(expiresAt)
      return formatted === '—' ? '会员已到期' : `已于 ${formatted} 到期`
    }
    case 'suspended':
      return '会员已暂停，请联系支持'
    case 'revoked':
      return '会员资格已撤销'
    case 'unknown':
      return '会员状态待确认'
    default:
      return ''
  }
}

const isUsageCount = (value: unknown): value is number =>
  typeof value === 'number' &&
  Number.isSafeInteger(value) &&
  value >= 0

const periodUsageIsAuthoritative = (subscription: QuotaSubscriptionDTO) => {
  const usage = subscription.period_usage
  if (
    usage?.state !== 'available' ||
    !isUsageCount(usage.total_requests) ||
    !isUsageCount(usage.total_tokens) ||
    !Array.isArray(usage.points)
  ) {
    return false
  }

  let requestTotal = 0
  let tokenTotal = 0
  for (const point of usage.points) {
    if (point.state === 'future') {
      if (
        point.requests != null ||
        point.cache_hit_tokens != null ||
        point.cache_miss_tokens != null ||
        point.output_tokens != null ||
        point.total_tokens != null
      ) {
        return false
      }
      continue
    }
    if (
      (point.state !== 'complete' && point.state !== 'partial') ||
      !isUsageCount(point.requests) ||
      !isUsageCount(point.cache_hit_tokens) ||
      !isUsageCount(point.cache_miss_tokens) ||
      !isUsageCount(point.output_tokens) ||
      !isUsageCount(point.total_tokens) ||
      point.total_tokens !==
        point.cache_hit_tokens + point.cache_miss_tokens + point.output_tokens
    ) {
      return false
    }
    requestTotal += point.requests
    tokenTotal += point.total_tokens
  }
  return (
    requestTotal === usage.total_requests &&
    tokenTotal === usage.total_tokens
  )
}

const mapUsagePoints = (
  subscription: QuotaSubscriptionDTO,
  usageAvailable: boolean
): DailyUsagePoint[] => {
  if (!usageAvailable) return []
  return (subscription.period_usage?.points ?? [])
    .filter(
      (point): point is typeof point & { state: 'complete' | 'partial' } =>
        point.state === 'complete' || point.state === 'partial'
    )
    .map((point) => ({
      date: point.start_at,
      label: formatDateTime(point.start_at).split(' ')[0],
      state: point.state,
      amount: 0,
      totalTokens: point.total_tokens as number,
      cacheHitTokens: point.cache_hit_tokens as number,
      cacheMissTokens: point.cache_miss_tokens as number,
      outputTokens: point.output_tokens as number,
      requests: point.requests as number
    }))
}

const timestamp = (value: string) => {
  const parsed = new Date(value).getTime()
  return Number.isFinite(parsed) ? parsed : Number.NEGATIVE_INFINITY
}

const isCurrentSubscription = (
  subscription: QuotaSubscriptionDTO,
  asOf: number
) => {
  if (subscription.status !== 'active' || !Number.isFinite(asOf)) return false
  const startsAt = timestamp(subscription.starts_at)
  const expiresAt = timestamp(subscription.expires_at)
  return (
    Number.isFinite(startsAt) &&
    Number.isFinite(expiresAt) &&
    startsAt <= asOf &&
    expiresAt > asOf
  )
}

const compareSubscriptionIdsDescending = (left: string, right: string) => {
  if (/^\d+$/.test(left) && /^\d+$/.test(right)) {
    const leftId = BigInt(left)
    const rightId = BigInt(right)
    if (leftId !== rightId) return leftId > rightId ? -1 : 1
  }
  return right.localeCompare(left)
}

const newestCurrentSubscriptionFirst = (
  subscriptions: QuotaSubscriptionDTO[],
  asOfValue: string
) => {
  const asOf = timestamp(asOfValue)
  return [...subscriptions].sort((left, right) => {
    const currentDifference =
      Number(isCurrentSubscription(right, asOf)) -
      Number(isCurrentSubscription(left, asOf))
    if (currentDifference !== 0) return currentDifference

    const startDifference =
      timestamp(right.starts_at) - timestamp(left.starts_at)
    if (startDifference !== 0) return startDifference

    const expiryDifference =
      timestamp(right.expires_at) - timestamp(left.expires_at)
    if (expiryDifference !== 0) return expiryDifference

    return compareSubscriptionIdsDescending(left.id, right.id)
  })
}

const mapSubscription = (
  subscription: QuotaSubscriptionDTO,
  asOf: number
): QuotaItem => {
  const window = subscription.weekly_window
  const monthlyWindow = subscription.monthly_window
  const monthlyUsedPercent = monthlyWindow?.used_percent
  const membershipStatus = mapMembershipStatus(subscription.status)
  const isActive = membershipStatus === 'active'
  const expiresAt = validTimestampOrNull(subscription.expires_at)
  const state = mapState(subscription)
  const resetsAt =
    isActive && state !== 'unknown'
      ? validTimestampOrNull(window.resets_at)
      : null
  const usageAvailable = periodUsageIsAuthoritative(subscription)
  const statusDetailLabel = membershipStatusDetail(
    membershipStatus,
    expiresAt
  )
  return {
    id: subscription.id,
    name: subscription.name,
    kindLabel: '订阅会员',
    periodLabel: '本周期',
    membershipStatus,
    statusDetailLabel,
    expiresAt,
    isCurrentMembership: isCurrentSubscription(subscription, asOf),
    state,
    usedPercent:
      window.state === 'unknown' ||
      typeof window.used_percent !== 'number' ||
      !Number.isFinite(window.used_percent)
        ? null
        : Math.max(0, Math.min(100, window.used_percent)),
    usedLabel: window.used == null ? '已用 —' : `已用 ${formatSnow(window.used)}`,
    limitLabel: window.limit == null ? '额度 —' : `额度 ${formatSnow(window.limit)}`,
    remainingLabel:
      window.remaining == null ? '剩余 —' : `剩余 ${formatSnow(window.remaining)}`,
    resetLabel:
      isActive
        ? resetsAt == null
          ? '重置时间待确认'
          : `${formatDateTime(resetsAt)} 重置`
        : statusDetailLabel,
    resetsAt,
    monthlyRemainingPercent:
      monthlyWindow?.state !== 'unknown' &&
      typeof monthlyUsedPercent === 'number' &&
      Number.isFinite(monthlyUsedPercent)
        ? Math.max(0, Math.min(100, 100 - monthlyUsedPercent))
        : null,
    periodStartLabel:
      isActive && window.period_start ? formatDateTime(window.period_start) : null,
    periodEndLabel:
      isActive && window.period_end ? formatDateTime(window.period_end) : null,
    tone: 'violet',
    icon: 'crown',
    usageAvailable,
    totalTokens: usageAvailable
      ? (subscription.period_usage?.total_tokens ?? null)
      : null,
    totalRequests: usageAvailable
      ? (subscription.period_usage?.total_requests ?? null)
      : null,
    points: mapUsagePoints(subscription, usageAvailable)
  }
}

export const adaptQuotaOverview = (input: unknown): QuotaOverview => {
  const dto = input as QuotaOverviewDTO
  if (
    !dto ||
    dto.schema_version !== 1 ||
    !dto.account ||
    !dto.wallet ||
    !Array.isArray(dto.subscriptions)
  ) {
    throw new Error('额度快照格式不受支持')
  }
  const asOf = timestamp(dto.as_of)
  return {
    accountName: dto.account.display_label || '落雪账户',
    accountState: dto.account.quota_state ?? 'unknown',
    dataScopeLabel:
      dto.account.data_scope === 'all_enabled_api_keys' ? '全部 API Key' : '账户额度',
    currency: dto.wallet.unit,
    balance: typeof dto.wallet.available === 'string' ? dto.wallet.available : null,
    todaySpend:
      typeof dto.wallet.today_spend === 'string' ? dto.wallet.today_spend : null,
    monthSpend:
      typeof dto.wallet.month_spend === 'string' ? dto.wallet.month_spend : null,
    reservedBalance:
      typeof dto.wallet.reserved === 'string' ? dto.wallet.reserved : null,
    balanceKeyCount: Number.isFinite(dto.wallet.balance_billed_key_count)
      ? dto.wallet.balance_billed_key_count
      : 0,
    walletState: dto.wallet.state ?? 'unknown',
    updatedAt: formatUpdatedAt(dto.generated_at),
    generatedAt: dto.generated_at,
    freshUntil: dto.fresh_until,
    quotas: newestCurrentSubscriptionFirst(dto.subscriptions, dto.as_of).map(
      (subscription) => mapSubscription(subscription, asOf)
    )
  }
}

File: backend/internal/handler/dto/quota_overview.go
package dto

import (
	"time"

	"github.com/Wei-Shaw/sub2api/internal/service"
)

type QuotaOverview struct {
	SchemaVersion   int                         `json:"schema_version"`
	RequestID       string                      `json:"request_id"`
	GeneratedAt     time.Time                   `json:"generated_at"`
	AsOf            time.Time                   `json:"as_of"`
	FreshUntil      time.Time                   `json:"fresh_until"`
	DisplayTimezone string                      `json:"display_timezone"`
	Freshness       string                      `json:"freshness"`
	Coverage        QuotaOverviewCoverage       `json:"coverage"`
	Account         QuotaOverviewAccount        `json:"account"`
	Wallet          QuotaOverviewWallet         `json:"wallet"`
	BillingGroups   []QuotaOverviewBillingGroup `json:"billing_groups"`
	Subscriptions   []QuotaOverviewSubscription `json:"subscriptions"`
	Actions         QuotaOverviewActions        `json:"actions"`
	Warnings        []string                    `json:"warnings"`
}

type QuotaOverviewCoverage struct {
	Included []string `json:"included"`
	Excluded []string `json:"excluded"`
}

type QuotaOverviewAccount struct {
	DisplayLabel      string              `json:"display_label"`
	DataScope         string              `json:"data_scope"`
	QuotaState        string              `json:"quota_state"`
	CanMakeRequest    *bool               `json:"can_make_request"`
	UsableGroupCount  int                 `json:"usable_group_count"`
	BlockedGroupCount int                 `json:"blocked_group_count"`
	UnknownGroupCount int                 `json:"unknown_group_count"`
	PrimaryIssue      *QuotaOverviewIssue `json:"primary_issue"`
}

type QuotaOverviewIssue struct {
	ScopeType         string     `json:"scope_type"`
	ScopeID           string     `json:"scope_id"`
	ReasonCode        string     `json:"reason_code"`
	RecommendedAction string     `json:"recommended_action"`
	RecoversAt        *time.Time `json:"recovers_at"`
}

type QuotaOverviewWallet struct {
	Unit                  string `json:"unit"`
	State                 string `json:"state"`
	Available             string `json:"available"`
	Reserved              string `json:"reserved"`
	TodaySpend            string `json:"today_spend"`
	MonthSpend            string `json:"month_spend"`
	BalanceBilledKeyCount int    `json:"balance_billed_key_count"`
}

type QuotaOverviewBillingGroup struct {
	ID                string                   `json:"id"`
	DisplayName       string                   `json:"display_name"`
	BillingMode       string                   `json:"billing_mode"`
	State             string                   `json:"state"`
	ReasonCode        *string                  `json:"reason_code"`
	RecommendedAction string                   `json:"recommended_action"`
	FallbackPolicy    string                   `json:"fallback_policy"`
	ResourceRef       QuotaOverviewResourceRef `json:"resource_ref"`
	Keys              []QuotaOverviewKey       `json:"keys"`
}

type QuotaOverviewResourceRef struct {
	Kind string  `json:"kind"`
	ID   *string `json:"id"`
}

type QuotaOverviewKey struct {
	ID        string `json:"id"`
	Name      string `json:"name"`
	MaskedKey string `json:"masked_key"`
	State     string `json:"state"`
}

type QuotaOverviewSubscription struct {
	ID            string                     `json:"id"`
	GroupID       string                     `json:"group_id"`
	Name          string                     `json:"name"`
	Status        string                     `json:"status"`
	StartsAt      time.Time                  `json:"starts_at"`
	ExpiresAt     time.Time                  `json:"expires_at"`
	WeeklyWindow  QuotaOverviewWeeklyWindow  `json:"weekly_window"`
	MonthlyWindow QuotaOverviewMonthlyWindow `json:"monthly_window"`
	PeriodUsage   QuotaOverviewPeriodUsage   `json:"period_usage"`
	NextEvent     *QuotaOverviewNextEvent    `json:"next_event"`
}

type QuotaOverviewPeriodUsage struct {
	State         string                          `json:"state"`
	ObservedUntil *time.Time                      `json:"observed_until"`
	BucketKind    string                          `json:"bucket_kind"`
	TotalRequests *int64                          `json:"total_requests"`
	TotalTokens   *int64                          `json:"total_tokens"`
	Points        []QuotaOverviewPeriodUsagePoint `json:"points"`
}

type QuotaOverviewPeriodUsagePoint struct {
	Index           int       `json:"index"`
	StartAt         time.Time `json:"start_at"`
	EndAt           time.Time `json:"end_at"`
	State           string    `json:"state"`
	Requests        *int64    `json:"requests"`
	CacheHitTokens  *int64    `json:"cache_hit_tokens"`
	CacheMissTokens *int64    `json:"cache_miss_tokens"`
	OutputTokens    *int64    `json:"output_tokens"`
	TotalTokens     *int64    `json:"total_tokens"`
}

type QuotaOverviewWeeklyWindow struct {
	Kind        string     `json:"kind"`
	State       string     `json:"state"`
	AnchorAt    time.Time  `json:"anchor_at"`
	PeriodStart *time.Time `json:"period_start"`
	PeriodEnd   *time.Time `json:"period_end"`
	ResetsAt    *time.Time `json:"resets_at"`
	Limit       *string    `json:"limit"`
	Used        *string    `json:"used"`
	Remaining   *string    `json:"remaining"`
	UsedPercent *float64   `json:"used_percent"`
}

type QuotaOverviewMonthlyWindow struct {
	Kind        string     `json:"kind"`
	State       string     `json:"state"`
	AnchorAt    time.Time  `json:"anchor_at"`
	PeriodStart *time.Time `json:"period_start"`
	PeriodEnd   *time.Time `json:"period_end"`
	ResetsAt    *time.Time `json:"resets_at"`
	Limit       *string    `json:"limit"`
	Used        *string    `json:"used"`
	Remaining   *string    `json:"remaining"`
	UsedPercent *float64   `json:"used_percent"`
}

type QuotaOverviewNextEvent struct {
	Kind string    `json:"kind"`
	At   time.Time `json:"at"`
}

type QuotaOverviewActions struct {
	RechargeURL            string `json:"recharge_url"`
	ManageKeysURL          string `json:"manage_keys_url"`
	ManageSubscriptionsURL string `json:"manage_subscriptions_url"`
}

func QuotaOverviewFromService(in *service.QuotaOverview) *QuotaOverview {
	if in == nil {
		return nil
	}
	out := &QuotaOverview{
		SchemaVersion:   in.SchemaVersion,
		RequestID:       in.RequestID,
		GeneratedAt:     in.GeneratedAt,
		AsOf:            in.AsOf,
		FreshUntil:      in.FreshUntil,
		DisplayTimezone: in.DisplayTimezone,
		Freshness:       in.Freshness,
		Coverage: QuotaOverviewCoverage{
			Included: append([]string(nil), in.Coverage.Included...),
			Excluded: append([]string(nil), in.Coverage.Excluded...),
		},
		Account: QuotaOverviewAccount{
			DisplayLabel:      in.Account.DisplayLabel,
			DataScope:         in.Account.DataScope,
			QuotaState:        in.Account.QuotaState,
			CanMakeRequest:    in.Account.CanMakeRequest,
			UsableGroupCount:  in.Account.UsableGroupCount,
			BlockedGroupCount: in.Account.BlockedGroupCount,
			UnknownGroupCount: in.Account.UnknownGroupCount,
			PrimaryIssue:      quotaOverviewIssueFromService(in.Account.PrimaryIssue),
		},
		Wallet: QuotaOverviewWallet{
			Unit:                  in.Wallet.Unit,
			State:                 in.Wallet.State,
			Available:             in.Wallet.Available,
			Reserved:              in.Wallet.Reserved,
			TodaySpend:            in.Wallet.TodaySpend,
			MonthSpend:            in.Wallet.MonthSpend,
			BalanceBilledKeyCount: in.Wallet.BalanceBilledKeyCount,
		},
		BillingGroups: make([]QuotaOverviewBillingGroup, 0, len(in.BillingGroups)),
		Subscriptions: make([]QuotaOverviewSubscription, 0, len(in.Subscriptions)),
		Actions: QuotaOverviewActions{
			RechargeURL:            in.Actions.RechargeURL,
			ManageKeysURL:          in.Actions.ManageKeysURL,
			ManageSubscriptionsURL: in.Actions.ManageSubscriptionsURL,
		},
		Warnings: append([]string(nil), in.Warnings...),
	}
	if out.Coverage.Included == nil {
		out.Coverage.Included = []string{}
	}
	if out.Coverage.Excluded == nil {
		out.Coverage.Excluded = []string{}
	}
	if out.Warnings == nil {
		out.Warnings = []string{}
	}
	for i := range in.BillingGroups {
		group := in.BillingGroups[i]
		mapped := QuotaOverviewBillingGroup{
			ID:                group.ID,
			DisplayName:       group.DisplayName,
			BillingMode:       group.BillingMode,
			State:             group.State,
			ReasonCode:        group.ReasonCode,
			RecommendedAction: group.RecommendedAction,
			FallbackPolicy:    group.FallbackPolicy,
			ResourceRef: QuotaOverviewResourceRef{
				Kind: group.ResourceRef.Kind,
				ID:   group.ResourceRef.ID,
			},
			Keys: make([]QuotaOverviewKey, 0, len(group.Keys)),
		}
		for j := range group.Keys {
			mapped.Keys = append(mapped.Keys, QuotaOverviewKey{
				ID:        group.Keys[j].ID,
				Name:      group.Keys[j].Name,
				MaskedKey: group.Keys[j].MaskedKey,
				State:     group.Keys[j].State,
			})
		}
		out.BillingGroups = append(out.BillingGroups, mapped)
	}
	for i := range in.Subscriptions {
		sub := in.Subscriptions[i]
		mapped := QuotaOverviewSubscription{
			ID:        sub.ID,
			GroupID:   sub.GroupID,
			Name:      sub.Name,
			Status:    sub.Status,
			StartsAt:  sub.StartsAt,
			ExpiresAt: sub.ExpiresAt,
			WeeklyWindow: QuotaOverviewWeeklyWindow{
				Kind:        sub.WeeklyWindow.Kind,
				State:       sub.WeeklyWindow.State,
				AnchorAt:    sub.WeeklyWindow.AnchorAt,
				PeriodStart: sub.WeeklyWindow.PeriodStart,
				PeriodEnd:   sub.WeeklyWindow.PeriodEnd,
				ResetsAt:    sub.WeeklyWindow.ResetsAt,
				Limit:       sub.WeeklyWindow.Limit,
				Used:        sub.WeeklyWindow.Used,
				Remaining:   sub.WeeklyWindow.Remaining,
				UsedPercent: sub.WeeklyWindow.UsedPercent,
			},
			MonthlyWindow: QuotaOverviewMonthlyWindow{
				Kind:        sub.MonthlyWindow.Kind,
				State:       sub.MonthlyWindow.State,
				AnchorAt:    sub.MonthlyWindow.AnchorAt,
				PeriodStart: sub.MonthlyWindow.PeriodStart,
				PeriodEnd:   sub.MonthlyWindow.PeriodEnd,
				ResetsAt:    sub.MonthlyWindow.ResetsAt,
				Limit:       sub.MonthlyWindow.Limit,
				Used:        sub.MonthlyWindow.Used,
				Remaining:   sub.MonthlyWindow.Remaining,
				UsedPercent: sub.MonthlyWindow.UsedPercent,
			},
			PeriodUsage: QuotaOverviewPeriodUsage{
				State:         sub.PeriodUsage.State,
				ObservedUntil: sub.PeriodUsage.ObservedUntil,
				BucketKind:    sub.PeriodUsage.BucketKind,
				TotalRequests: sub.PeriodUsage.TotalRequests,
				TotalTokens:   sub.PeriodUsage.TotalTokens,
				Points: make(
					[]QuotaOverviewPeriodUsagePoint,
					0,
					len(sub.PeriodUsage.Points),
				),
			},
		}
		if sub.PeriodUsage.Points == nil {
			mapped.PeriodUsage.Points = nil
		} else {
			for j := range sub.PeriodUsage.Points {
				point := sub.PeriodUsage.Points[j]
				mapped.PeriodUsage.Points = append(
					mapped.PeriodUsage.Points,
					QuotaOverviewPeriodUsagePoint{
						Index:           point.Index,
						StartAt:         point.StartAt,
						EndAt:           point.EndAt,
						State:           point.State,
						Requests:        point.Requests,
						CacheHitTokens:  point.CacheHitTokens,
						CacheMissTokens: point.CacheMissTokens,
						OutputTokens:    point.OutputTokens,
						TotalTokens:     point.TotalTokens,
					},
				)
			}
		}
		if sub.NextEvent != nil {
			mapped.NextEvent = &QuotaOverviewNextEvent{
				Kind: sub.NextEvent.Kind,
				At:   sub.NextEvent.At,
			}
		}
		out.Subscriptions = append(out.Subscriptions, mapped)
	}
	return out
}

func quotaOverviewIssueFromService(in *service.QuotaOverviewIssue) *QuotaOverviewIssue {
	if in == nil {
		return nil
	}
	return &QuotaOverviewIssue{
		ScopeType:         in.ScopeType,
		ScopeID:           in.ScopeID,
		ReasonCode:        in.ReasonCode,
		RecommendedAction: in.RecommendedAction,
		RecoversAt:        in.RecoversAt,
	}
}
